CMakeLists.txt

```diff
@@ -47,7 +47,7 @@ add_subdirectory(3rd-party)
 find_package(Qt5 COMPONENTS Core Network Widgets REQUIRED)
 
 function(determine_version OUTPUT_VARIABLE)
-  execute_process(COMMAND git describe --all --match release/*
+  execute_process(COMMAND git describe --all --match origin/release/*
```

Comment:
Could you please clarify what cases were malfunctioning without this?

Reply:
This only matched `release/0.9` if you had it as a local branch (i.e. `heads/release/0.9`). The new match matches on the `origin` remote (i.e. `remotes/origin/release/0.9`), so you can check out the commit alone (i.e. `git fetch origin/release/0.9; git checkout FETCH_HEAD`).

That said, I _think_ we're using newer git now when building the snap, so I should be able to use a more precise regex here instead. Will circle back.

Reply:
Yep, that's clear. I was more asking if there was something in CI or some real use-case that requires this. I am concerned that
- this assumes an `origin` remote. For instance, I don't have one with that name here (I call it `upstream`). And I could actually have a remote called `origin` pointing to some other repo (e.g. a personal fork). One way to address this would be `git describe --all --match *release/0.9`, but then there could be multiple/different remotes...
- determining the version depends on something external. It feels like it should be determined according to the repository this is run in. So, no branch named `release/0.9` there &rarr; not version 0.9 in that repo. To determine an official version, this would just need to run in a representative repository, with the required branches and tags.

On a side note, `git fetch origin/release/0.9` does not work for me, git says it "does not appear to be a git repository". But this can be tested under a branch with a different name (e.g. `git checkout -b special_branch upstream/release/0.9`).

Reply:
> I was more asking if there was something in CI or some real use-case that requires this

Yes, CI gets told to build a particular commit (because the referencing branch might've moved since the job was triggered).

>     * this assumes an `origin` remote. For instance, I don't have one with that name here (I call it `upstream`). And I could actually have a remote called `origin` pointing to some other repo (e.g. a personal fork).

Sure, but that's the default remote name, so not a crazy assumption. And where it really matters (CI), we're in control of the remote name.

> One way to address this would be `git describe --all --match *release/0.9`, but then there could be multiple/different remotes...

Yeah, that.

>     * determining the version depends on something external. It feels like it should be determined according to the repository this is run in. So, no branch named `release/0.9` there → not version 0.9 in that repo. To determine an official version, this would just need to run in a representative repository, with the required branches and tags.

That's exactly the problem, the context we get in CI is just the commit, and I'm arguing here that authority of whether that commit is a release candidate (or a release, for that matter) is the upstream repository (by means of a branch name, or a tag), not the local one. Of course you can spoof that by overwriting the tag or the remote, if you so wish.

An alternative is to use a CMake variable and be explicit about what constitutes a release canidate. But then it becomes a function of _how_, rather than _what_ is being built.

> On a side note, `git fetch origin/release/0.9` does not work for me, git says it "does not appear to be a git repository". But this can be tested under a branch with a different name (e.g. `git checkout -b special_branch upstream/release/0.9`).

This should've been `git fetch origin release/0.9`, sorry. And it's different to what you did in that there is no branch if you `git checkout FETCH_HEAD` - you end up in _detached_ state.

Reply:
> Ahh I see. OK, I understand it better now, I agree we need to depend on the remote then. I'd be happier with the star version then. A different remote config would affect the output, but I guess it then makes sense, and then it wouldn't depend on remote names. WDYT?

> Yeah that the command? Or yeah that there could be different remotes?

More importantly, `*` also matches `/`, so a local "foorelease/0.9" branch would also match. Which is why I'm looking at the regular expression approach (that's what I started with, but couldn't use due to older git on Core 16).

Different remotes, partial matches… Wildcards are wild.

Reply:
> Which is why I'm looking at the regular expression approach

Ugh, there isn't a regular expression approach, it's only globs. I misremembered - the difference was multiple `--match` support, not regexes.

I'd go for being explicit about `origin/` in the match, as even this:
```
git describe --all --match */release/*
```
Can be fooled by a local `foo/release/0.9` branch… It's meh that I can't do `--match refs/remotes/*/release/0.9` - that would at least limit it to remotes…

Reply:
A `foo/release/0.9` branch sounds less likely than an absent (or fork-pointing) `origin`. I agree the ideal would be a regex, but failing that, IMO it makes more sense to impose restrictions on branches than on remotes.

We could rule out branches with slashes except `release/<num>` and error out in the regex when the glob pattern you mentioned (`*/release/*`) caught anything before "release" that was not a remote.

Orrr... we use `${AUTHORITATIVE_REMOTE_PREFIX}release/*` and pass that var in CI. Perhaps this would be the most sensible... WDYT?

Reply:
Yeah, how about a40e7a65f41b25d12ef3df3d6dfc251a9a1d844e?

Reply:
That is a bit better, but now I need to remember to setup an environment or I get the wrong version. Don't you think it would make more sense to use local stuff and override in CI? The detached HEAD would be the special situation needing the var, and that would be in code (rather than local copies needing manual env setup).

Reply:
Yeah, I think that is actually better than getting the wrong version unless you have the right remote. I will ask you to reconsider. Proposal in #1160 (using this as base).

That way working copies are more independent by default, as until now. But we can give them an overriding reference.

Reply:
BTW, shouldn't the remote be used in other `describe` commands too?

Reply:
Hey @ricab, I've incorporated your proposal and tweaked it a bit:
- when `HEAD` exactly matches `release/*`, `MULTIPASS_UPSTREAM` is required
- you need to set `MULTIPASS_UPSTREAM=""` to match in the local repo